### PR TITLE
Add BarDistribution class and update TabDPTRegressor for a ScoringBench compatible API

### PR DESCRIPTION
**Pull request by ikozl090**
Introduced the BarDistribution class for handling distributions over binned continuous targets. Updated TabDPTRegressor to utilize BarDistribution for regression outputs, including mean, median, mode, and quantiles. Enhanced output handling with new output types and improved initialization for constant target scenarios. Updated __all__ exports in __init__.py to include new output types.

### CODE REVIEW

File: src/tabdpt/bar_distribution.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+
+class BarDistribution(nn.Module):
+    """Distribution over fixed-width bins defined by border edges."""
+
+    def __init__(self, borders: torch.Tensor):
+        super().__init__()
+        borders = borders.contiguous()
+        self.register_buffer("borders", borders)
+        assert (self.bucket_widths >= 0.0).all(), "borders must be sorted"
```
Comment (anthonycaterini) [inline src/tabdpt/bar_distribution.py]: what happens if a bucket has width 0?

File: src/tabdpt/bar_distribution.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        left_border = self.borders[idx]
+        right_border = self.borders[idx + 1]
+        return left_border + (right_border - left_border) * rest_prob / probs.gather(
+            -1,
+            idx[..., None],
+        ).squeeze(-1)
+
+    def mode(self, logits: torch.Tensor) -> torch.Tensor:
+        density = logits.softmax(dim=-1) / self.bucket_widths
+        mode_inds = density.argmax(dim=-1)
```
Comment (anthonycaterini) [inline src/tabdpt/bar_distribution.py:139]: what happens if there are multiple modes?

File: src/tabdpt/bar_distribution.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+"""Distribution over binned (bucketed) continuous targets."""
+
+from __future__ import annotations
+
+import torch
+from torch import nn
+
+
+class BarDistribution(nn.Module):
+    """Distribution over fixed-width bins defined by border edges."""
```
Comment (anthonycaterini) [inline src/tabdpt/bar_distribution.py]: Helpful things for this docstring would be the typical shapes of:

1. `borders`
2. `logits`

File: src/tabdpt/bar_distribution.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            idx[..., None],
+        ).squeeze(-1)
+
+    def mode(self, logits: torch.Tensor) -> torch.Tensor:
+        density = logits.softmax(dim=-1) / self.bucket_widths
+        mode_inds = density.argmax(dim=-1)
+        bucket_means = self.borders[:-1] + self.bucket_widths / 2
+        return bucket_means[mode_inds]
+
+    def sample(self, logits: torch.Tensor) -> torch.Tensor:
```
Comment (anthonycaterini) [inline src/tabdpt/bar_distribution.py:143]: Some comments on this function:

1. Should add a docstring to make the input and output shapes more clear. I'm confused about what exactly is being sampled - assume it's one sample per batch dimension, but again it's not super clear.
2. Relatedly, what is the shape of `u`? it's `*logits.shape[:-1]`? Why not just `logits.shape[0]`, i.e., the batch dimension?
3. Also related to point 1, what if people want to sample `n` times for each `logit`? Should we enable that capability?
4. `self.icdf` claims to be taking a `float` for `left_prob`, but this is a `torch.tensor`.

File: src/tabdpt/regressor.py [hunk 5]
```diff
@@ -119,36 +188,170 @@ def _predict(
...
         for inner_seed in self._get_ensemble_iterator(n_ensembles, seed):
-            inner_seed = int(inner_seed)
-            prediction_cumsum += self._predict(X, context_size=context_size, batch_size=batch_size, seed=inner_seed)
+            logits = self._predict_logits(
+                X,
+                context_size=context_size,
+                batch_size=batch_size,
+                seed=int(inner_seed),
+            )
+            accumulated = logits if accumulated is None else accumulated + logits
```
Comment (anthonycaterini) [inline src/tabdpt/regressor.py]: why not just start `logits` at zero?

File: src/tabdpt/regressor.py [hunk 6]
```diff
@@ -119,36 +188,170 @@ def _predict(
...
+
+        The predictive distribution is a fixed-bin histogram over z-normalized targets
+        in ``[regression_bin_min, regression_bin_max]``, mapped to raw units via
+        fit-time ``y_train_mean_`` and ``y_train_std_``. All probability mass lies within
+        approximately ``[mean - 10*std, mean + 10*std]`` in raw space. Quantile bands
+        reflect the model's binned predictive distribution, not guaranteed coverage
+        intervals.
+        """
+        if quantiles is None:
+            quantiles = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]
```
Comment (anthonycaterini) [inline src/tabdpt/regressor.py]: why do we need to specify `quantiles` if we are doing mean, median, mode, etc?

File: tests/reg_example.py
Comment (anthonycaterini) [inline tests/reg_example.py:1]: this file actually feels now too complicated for the simple example it was meant to be. Suggest taking the innovations in this file into a more involved `reg_example.py` (and updating the `README` accordingly)

**Review by anthonycaterini — CHANGES_REQUESTED** [review]
Some small changes and clarifications requested

Another idea: it would be nice to have a regression test (pun not intended) to check for model changes stemming from changes to the regression code. In theory, the changes in here should give the same result as before, unless seeds and randomness get weird

File: src/tabdpt/bar_distribution.py (hunk 1 above)
Comment (ikozl090) [reply src/tabdpt/bar_distribution.py]: Good point, changed the check from `>=` to `>`!

File: src/tabdpt/bar_distribution.py (hunk 2 above)
Comment (ikozl090) [reply src/tabdpt/bar_distribution.py:139]: Do you mean if there are a few bins with equal probabilities?

Comment (ikozl090) [reply src/tabdpt/bar_distribution.py:139]: If so, then it will probably have to pick one of the bars with equal densities in some arbitrary way, but it does seem to be the way that TabPFN implemented this too so maybe the probability of having two buckets with exactly equal densities is just very low so it's rarely an issue?

File: src/tabdpt/bar_distribution.py (hunk 4 above)
Comment (ikozl090) [reply src/tabdpt/bar_distribution.py:143]: 1. Sounds good, adding more comments and a docstring! 
2. It was meant to be a more general way of dealing with inputs of all sort of dimensions, for instance if we use retrieval would the input shape be something like (batch_dim, seq_leq)? I know the seq length should be 1 then, but it still seems like a more natural way to make sure the input and output shape match no matter what.
3. I don't think it's something that TabPFN is implementing right now, but if you think it's worth it, it shouldn't be an issue to enable. This can always be done separately by the user by just outputting the logits also, do you think this would be a common enough use case to require enabling natively? 
4. You're right, this was an error! It was meant to be compatible with both tensors and floats, updating!

File: src/tabdpt/bar_distribution.py (hunk 2 above)
Comment (anthonycaterini) [reply src/tabdpt/bar_distribution.py:139]: yes this is what I meant. maybe it was worth writing a note in the code somewhere about this behaviour

File: src/tabdpt/regressor.py (hunk 5 above)
Comment (ikozl090) [reply src/tabdpt/regressor.py]: Fixed this! There wasn't any reason to change this part.

File: src/tabdpt/regressor.py (hunk 6 above)
Comment (ikozl090) [reply src/tabdpt/regressor.py]: Moved this so default values are only specified when one of the output modes that require quantiles is used. This would hopefully make it a bit clearer when they are and aren't necessary.

File: src/tabdpt/bar_distribution.py (hunk 2 above)
Comment (ikozl090) [reply src/tabdpt/bar_distribution.py:139]: Added quick comment on argmax behavior to clarify.

File: src/tabdpt/bar_distribution.py (hunk 4 above)
Comment (anthonycaterini) [reply src/tabdpt/bar_distribution.py:143]: probably fine to not have multiple samples at once

**Comment by anthonycaterini** [comment]
My comments are resolved, except for one new one: It is mentioned that the shape of things is now `(*batch, num_bars)`, but is it correct to lump everything in with `batch` when many of the dimensions are not the batch dimension?

**Comment by ikozl090** [comment]
> My comments are resolved, except for one new one: It is mentioned that the shape of things is now `(*batch, num_bars)`, but is it correct to lump everything in with `batch` when many of the dimensions are not the batch dimension?

Good point, I modified the language for these dimensions to be a bit more general. Let me know if it makes more sense now!

**Review by anthonycaterini — APPROVED** [review]

File: src/tabdpt/regressor.py [hunk 7]
```diff
@@ -119,36 +192,162 @@ def _predict(
...
+        self,
+        X: np.ndarray,
+        *,
+        output_type: OutputType = "mean",
+        quantiles: list[float] | None = None,
+        n_ensembles: int = 8,
+        context_size: int | None = None,
+        batch_size: int | None = None,
+        seed: int | None = None,
+    ) -> RegressionResultType:
```
Comment (j201) [inline src/tabdpt/regressor.py]: Personally, I would prefer to just have a single type definition with a type union output rather than spending so much space on being more specific than that

**Comment by anthonycaterini** [comment]
oh btw @ikozl090 we should normalize the backticks in the docstrings of this PR with the rest of the codebase, i.e., just use one and not two delimiters on each side

**Review by j201 — COMMENTED** [review]
Thanks, I think this a lot cleaner and more maintainable in terms of the changes to `regressor.py`. Just a couple questions:

- What's the thinking behind having `mean`, `median`, and `mode` as methods of `BarDistribution` but having the corresponding numpy output functions be standalone functions? My immediate inclination would actually be to just have a set of functions that take in a `FullPrediction` for all the cases where the logits and bar distribution are needed, but I don't feel that strongly about it.
- Why is `BarDistribution` an `nn.Module` subclass?

Minor points:
- We've generally been using a line length of 120, not 80, but I don't think you need to go back and change that
- In #77 we'll be moving examples to the `examples/` folder, so either that PR or this one should make sure to move the `reg_probabilistic_example.py` file.

**Comment by ikozl090** [comment]
> Thanks, I think this a lot cleaner and more maintainable in terms of the changes to `regressor.py`. Just a couple questions:
> 
> * What's the thinking behind having `mean`, `median`, and `mode` as methods of `BarDistribution` but having the corresponding numpy output functions be standalone functions? My immediate inclination would actually be to just have a set of functions that take in a `FullPrediction` for all the cases where the logits and bar distribution are needed, but I don't feel that strongly about it.
> * Why is `BarDistribution` an `nn.Module` subclass?
> 
> Minor points:
> 
> * We've generally been using a line length of 120, not 80, but I don't think you need to go back and change that
> * In #77 we'll be moving examples to the `examples/` folder, so either that PR or this one should make sure to move the `reg_probabilistic_example.py` file.

Agreed on the comments, the current structure was still trying to be somewhat aligned with the TabPFN output structure (which is why BarDistribution was an `nn.Module`), but I don't think that's necessary for the current repo structure. I dropped BarDistribution entirely. predict(..., output_type="full") now just returns a FullPrediction object with logits and bin borders, and the numpy stats are standalone helpers that take that object.

Also aligned the new code to 120 chars and moved the example to examples/reg_probabilistic_example.py so it won’t get left behind by #77.

**Review by j201 — APPROVED** [review]
